hudi-spark/src/main/java/org/apache/hudi/keygen/CustomKeyGenerator.java

```diff
@@ -0,0 +1,141 @@
+  private static final String DEFAULT_PARTITION_PATH_SEPARATOR = "/";
+  private static final String SPLIT_REGEX = ":";
+
+  public CustomKeyGenerator(TypedProperties props) {
+    super(props);
+    this.properties = props;
+    this.recordKeyFields = Arrays.stream(props.getString(DataSourceWriteOptions.RECORDKEY_FIELD_OPT_KEY()).split(",")).map(String::trim).collect(Collectors.toList());
+    this.partitionPathFields =
+      Arrays.stream(props.getString(DataSourceWriteOptions.PARTITIONPATH_FIELD_OPT_KEY()).split(",")).map(String::trim).collect(Collectors.toList());
+  }
+
+  @Override
+  public HoodieKey getKey(GenericRecord record) {
+    //call function to get the record key
+    String recordKey = getRecordKey(record);
+    //call function to get the partition key based on the type for that partition path field
+    String partitionPath = getPartitionPath(record);
+    return new HoodieKey(recordKey, partitionPath);
+  }
+
+  public String getPartitionPath(GenericRecord record) {
+    if (partitionPathFields == null || partitionPathFields.isEmpty()) {
+      throw new HoodieKeyException("Unable to find field names for partition path in cfg");
+    }
+
+    String partitionPathField;
+    StringBuilder partitionPath = new StringBuilder();
+    boolean nonPartitionedTable = false;
+    for (String field : partitionPathFields) {
+      String[] fieldWithType = field.split(SPLIT_REGEX);
+      if (fieldWithType.length != 2) {
+        throw new HoodieKeyException("Unable to find field names for partition path in proper format");
+      }
+
+      partitionPathField = fieldWithType[0];
+      switch (getPartitionKeyType(fieldWithType[1])) {
+        case SIMPLE:
+          SimpleKeyGenerator simpleKeyGenerator = new SimpleKeyGenerator(properties);
+          partitionPath.append(simpleKeyGenerator.getPartitionPath(record, partitionPathField));
+          break;
+        case TIMESTAMP_BASED:
+          TimestampBasedKeyGenerator timestampBasedKeyGenerator = new TimestampBasedKeyGenerator(properties);
+          partitionPath.append(timestampBasedKeyGenerator.getPartitionPath(record, partitionPathField));
+          break;
+        case NOPARTITION:
+          NonpartitionedKeyGenerator nonpartitionedKeyGenerator = new NonpartitionedKeyGenerator(properties);
+          partitionPath.append(nonpartitionedKeyGenerator.getPartitionPath(record, partitionPathField));
+          nonPartitionedTable = true;
+          break;
+        default:
+          throw new HoodieDeltaStreamerException("Please provide valid PartitionKeyType with fields!");
+      }
+
+      if (nonPartitionedTable) {
+        return "";
+      }
+      partitionPath.append(DEFAULT_PARTITION_PATH_SEPARATOR);
+    }
+    partitionPath.deleteCharAt(partitionPath.length() - 1);
+
+    return partitionPath.toString();
+  }
+
+  private PartitionKeyType getPartitionKeyType(String type) {
+    switch (type) {
+      case "simple":
+        return SIMPLE;
+      case "timestampBased":
```

Comment:
we can can skip the "Based" ? just `timestamp` ?

Reply:
Ok. Done.